### PR TITLE
INC-739: PATCH /sync/booking/{bookingId}/id/{id} endpoint

### PR DESCRIPTION
**Pull request by xoen**
(**NOTE**: Probably easier to review commit by commit.)

Adds endpoint to update an existing IEP review (`PrisonerIepLevel`)

As per ticket:
- `MAINTAIN_IEP` role and write scope needed to perform this operation
- only `comment`, `iepTime` and `current` fields can be changed
- publishes relevant events:
  - `"incentives.iep-review.updated"` domain event ([domain event types are now part of the `IncentivesDomainEventType` enum](https://github.com/ministryofjustice/hmpps-incentives-api/commit/9db71410cb89fb318e6288fe38ff55ef7e8c9352), e.g. this endpoint publishes `IncentivesDomainEventType.IEP_REVIEW_UPDATED`)
  - also publishes `AuditType.IEP_REVIEW_UPDATED` audit message
  - made existing code that publishes domain/audit events more explicit about what events are published
- revamped Swagger docs so that:
  - example requests bodies are useful (courtesy of `@Schema`'s `implementation` field)
  - example values are consistent across all the swagger examples, so that requests tend to match responses and they're less likely to confuse people
  - I guess this was technically part of [INC-740](https://dsdmoj.atlassian.net/browse/INC-740) but I needed to have docs for this endpoint (...and I may have been carried away a little 😅)
- tests to endpoint and also to the corresponding service

### CODE REVIEW

**Comment by xoen** [comment]
FYI: I've [fixed the few linting problems](https://github.com/ministryofjustice/hmpps-incentives-api/pull/174/commits/2b16c89dca7d00915ef34e1d67de674eb2122f75) hopefully

File: src/test/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewServiceTest.kt [hunk 1]
```diff
@@ -364,6 +365,100 @@ class PrisonerIepLevelReviewServiceTest {
     }
   }
 
+  @Nested
+  inner class HandleSyncPatchIepReviewRequest {
+
+    // TODO: Tests to update
```
Comment (neilmendum) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewServiceTest.kt:371]: Think you can remove this `TODO`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewService.kt [hunk 2]
```diff
@@ -95,7 +97,42 @@ class PrisonerIepLevelReviewService(
...
+    return iepDetail
+  }
+
+  suspend fun handleSyncPatchIepReviewRequest(bookingId: Long, id: Long, syncPatchRequest: SyncPatchRequest): IepDetail {
+    if (listOf(syncPatchRequest.iepTime, syncPatchRequest.comment, syncPatchRequest.current).all { it == null }) {
+      throw ValidationException("Please provide fields to update")
+    }
+
+    var prisonerIepLevel: PrisonerIepLevel? = prisonerIepLevelRepository.findById(id)
+    if (prisonerIepLevel == null) {
```
Comment (neilmendum) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewService.kt:114]: You could convert this to `?: run {` but not a big one either way

**Review by neilmendum — COMMENTED** [review]
Looks great - 2 minor comments

File: src/test/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewServiceTest.kt (hunk 1 above)
Comment (xoen) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewServiceTest.kt]: Oh yes left over I forgot, I noticed while I was checking the flaky tests again, [removed here](https://github.com/ministryofjustice/hmpps-incentives-api/pull/174/commits/4ef476e28e202fdf33549bfeb359d94e7a7f82ad) ☺️

**Comment by xoen** [comment]
@Mjwillis FYI: This PR adds the sync `PATCH` endpoint. 

Regarding domain/audit events:
- [I've added an enum](https://github.com/ministryofjustice/hmpps-incentives-api/pull/174/commits/9db71410cb89fb318e6288fe38ff55ef7e8c9352#diff-fad93080a1e69c7198e25624f1befabbb6377e9b2767e15649c0cfe6884197fcR76-R79) for domain events published by this service `IncentivesDomainEventType` (new domain event type is `"incentives.iep-review.updated"`, named consistently with the existing `"incentives.iep-review.inserted"` one)
- [I've also added the `IEP_REVIEW_UPDATED` type](https://github.com/ministryofjustice/hmpps-incentives-api/pull/174/commits/9db71410cb89fb318e6288fe38ff55ef7e8c9352#diff-b71856ff05ded2bd0b5c690b45ca91d001b430614cc94ac35f9de4d718557bc2R67-R70) to the existing `AuditType`
- nothing too exciting but hopefully it's a bit tidier using enums for both events types

CC: @neilmendum

File: src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewService.kt (hunk 2 above)
Comment (xoen) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/incentivesapi/service/PrisonerIepLevelReviewService.kt:114]: As discussed over Slack I'll leave the `if` as it is.

I've played locally to use `?: run { }` but:
1. seems to cause subsequent lines to complain about the value possibly be `null` (which is weird as I'm throwing an exception and Kotlin should be able to infer that the specific type is a `PrisonerIepLevel` not a `PrisonerIepLevel?` anymore (like it does in the `if` case)
2. it doesn't seem very readable

:)

**Review by neilmendum — APPROVED** [review]
LGTM

**Pull request merged**